### PR TITLE
added sql db syncGroupPrivateLink

### PR DESCRIPTION
**Pull request by muzzamilinovaqo**

### CODE REVIEW

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'Private Link Enabled for SQL Databases',
```
Comment (fatima99s) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:5]:
```suggestion
    title: 'SQL Databases Private Link Enabled',
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'Database Private Link Enabled',
+    category: 'SQL Databases',
+    domain: 'Databases',
+    description: 'Ensures SQL Database sync groups use private link when SQL DB sync with others databases.',
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:8]:
```suggestion
    description: 'Ensures SQL Database sync groups are configured to use private link.',
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    more_info: 'Private Link is the way to create Sync groups using secure connection with databases sitting behind a firewall. SQL Data Sync Private Link is Microsoft-managed endpoint and internally creates a subnet within the existing virtual network, so there is no need to create another virtual network or subnet.',
+    recommended_action: 'Configure SQL Database sync groups to use private link and mandate manual approval for private endpoint connections.',
+    link: 'https://learn.microsoft.com/en-us/azure/private-link/private-link-overview',
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:11]:
```suggestion
    link: 'https://learn.microsoft.com/en-us/azure/azure-sql/database/sql-data-sync-data-sql-server-sql-database?view=azuresql',
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'Database Private Link Enabled',
+    category: 'SQL Databases',
+    domain: 'Databases',
+    description: 'Ensures SQL Database sync groups use private link when SQL DB sync with others databases.',
+    more_info: 'Private Link is the way to create Sync groups using secure connection with databases sitting behind a firewall. SQL Data Sync Private Link is Microsoft-managed endpoint and internally creates a subnet within the existing virtual network, so there is no need to create another virtual network or subnet.',
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:9]:
```suggestion
    more_info: 'Private link feature allows you to choose a service managed private endpoint to establish a secure connection between the sync service and your member/hub databases during the data synchronization process. A service managed private endpoint is a private IP address within a specific virtual network and subnet.',
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    recommended_action: 'Configure SQL Database sync groups to use private link and mandate manual approval for private endpoint connections.',
+    link: 'https://learn.microsoft.com/en-us/azure/private-link/private-link-overview',
+    apis: ['servers:listSql','databases:listByServer','syncGroups:list'],
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:12]:
```suggestion
    apis: ['servers:listSql', 'databases:listByServer', 'syncGroups:list'],
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+                if (!databases || databases.err || !databases.data) {
+                    helpers.addResult(results, 3,
+                        'Unable to query for SQL server databases: ' + helpers.addError(databases), location, server.id);
+                } else {
+                    if (!databases.data.length) {
+                        helpers.addResult(results, 0,
+                            'No databases found for SQL server', location, server.id);
+                    } else {
+                        databases.data.forEach(database=> {
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:48]:
```suggestion
                        databases.data.forEach(database => {
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                            
+                            var syncGroups = helpers.addSource(cache, source, ['syncGroups', 'list', location, database.id]);
+
+                            if (!syncGroups || syncGroups.err || !syncGroups.data) {
+                                helpers.addResult(results, 3, 'Unable to query for SQL Database sync groups: ' + helpers.addError(syncGroups), location, database.id);
+                                return;
+                            }
+                            if (!syncGroups.data.length) {
+                                helpers.addResult(results, 0,
+                                    'No Database sync group found for SQL database', location, database.id);
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:58]:
```suggestion
                                    'No sync groups found for SQL database', location, database.id);
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                                helpers.addResult(results, 0,
+                                    'No sync groups found for SQL database', location, database.id);
+                            }
+
+                            syncGroups.data.forEach(syncGroup=> {
+                                if (syncGroup.usePrivateLinkConnection) {
+                                    helpers.addResult(results, 0, 'Database sync group uses private link to sync with other databases', location, syncGroup.id);
+                                } else {
+                                    helpers.addResult(results, 2, 'Database sync group does not uses private link to sync with other databases', location, syncGroup.id);
+                                }
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:66]: Again the result should be SQL Based instead of syncGroup based

**Review by alphadev4 — CHANGES_REQUESTED** [review]

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                                return;
+                            }
+                            if (!syncGroups.data.length) {
+                                helpers.addResult(results, 0,
+                                    'No sync groups found for SQL database', location, database.id);
+                            } else {
+                                var missingPrivateConfigGrps = syncGroups.data.filter((e) => !e.usePrivateLinkConnection).map((e)=>e.name);
+    
+                                if (missingPrivateConfigGrps.length) {
+                                    helpers.addResult(results, 2, `Database is not configured to use private link in following sync groups: ${missingPrivateConfigGrps.join(', ')} `, location, database.id);
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:63]:
```suggestion
                                    helpers.addResult(results, 2, `SQL Database following sync groups are not configured to use private link: ${missingPrivateConfigGrps.join(', ')} `, location, database.id);
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+                                   
+                                } else {
+                                    helpers.addResult(results, 0, 'Database sync groups are configured to use private link', location, database.id);
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:66]:
```suggestion
                                    helpers.addResult(results, 0, 'SQL Database sync groups are configured to use private link', location, database.id);
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.spec.js [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+            );
+
+            sqlDatabaseSyncGroups.run(cache, {}, callback);
+        });
+
+        it('should give failing result if SQL Database sync group does not use private link', function(done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('Database is not configured to use private link in following sync groups:');
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.spec.js:118]:
```suggestion
                expect(results[0].message).to.include('SQL Database following sync groups are not configured to use private link:');
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.spec.js [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+            );
+
+            sqlDatabaseSyncGroups.run(cache, {}, callback);
+        });
+
+        it('should give passing result if SQL Database sync group uses private link', function(done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('Database sync groups are configured to use private link');
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.spec.js:100]:
```suggestion
                expect(results[0].message).to.include('SQL Database sync groups are configured to use private link');
```

**Review by alphadev4 — APPROVED** [review]

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+                            if (!syncGroups || syncGroups.err || !syncGroups.data) {
+                                helpers.addResult(results, 3, 'Unable to query for SQL Database sync groups: ' + helpers.addError(syncGroups), location, database.id);
+                                return;
+                            }
+                            if (!syncGroups.data.length) {
+                                helpers.addResult(results, 0,
+                                    'No sync groups found for SQL database', location, database.id);
+                            } else {
+                                var missingPrivateConfigGrps = syncGroups.data.filter((e) => !e.usePrivateLinkConnection).map((e)=>e.name);
```
Comment (mehakseedat63) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:60]:
```suggestion
                                var missingPrivateConfigGrps = syncGroups.data.filter((e) => !e.usePrivateLinkConnection).map((e) => e.name);
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    
+                                if (missingPrivateConfigGrps.length) {
+                                    helpers.addResult(results, 2, `SQL Database following sync groups are not configured to use private link: ${missingPrivateConfigGrps.join(', ')} `, location, database.id);
```
Comment (mehakseedat63) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:63]:
```suggestion
                                    helpers.addResult(results, 2, `Following SQL Database sync groups are not configured to use private link: ${missingPrivateConfigGrps.join(', ')} `, location, database.id);
```

File: plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+                                   
+                                } else {
+                                    helpers.addResult(results, 0, 'SQL Database sync groups are configured to use private link', location, database.id);
```
Comment (mehakseedat63) [inline plugins/azure/sqldatabases/dbSyncGroupPrivateLink.js:66]:
```suggestion
                                    helpers.addResult(results, 0, 'All SQL Database sync groups are configured to use private link', location, database.id);
```

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]

**Review by mehakseedat63 — APPROVED** [review]

**Pull request merged**